### ISSUE TITLE
Need help for using the inbuilt PD controller of MuJoCo

### ISSUE THREAD

**Issue by rishabkr29** [google-deepmind/mujoco#1337]
Hello MuJoCo Community, 

I'm a student  and I'm trying to use MuJoCo for thesis.

I'm looking for some help related to inbuilt PD controller of MuJoCo.

Here is a model which explains my question: based on the model I am calculating the slider displacement and directly passing it to the data.ctrl (which means I am trying to use feedforward as my high level ) and rest I am dependent on the low level pd of the mujoco (I read somewhere mujoco have inbuilt PD). 

My question is :
1.  My end effector of the system should  reach  to the refence value. Instead of giving the correct slider length and using the correct variable ids. why it is not able to achieve. 

<details>
  <summary>here is the controller section of Python file</summary>

**Comment by rishabkr29** [comment]
here is my xml model : `<mujoco>

  <option timestep="0.01" gravity="0 0 -9.81" integrator="Euler" jacobian="dense">
    <flag energy="enable" constraint = "enable" contact="enable" sensornoise="enable"/>   
  </option>

  

  <compiler autolimits="true"/>   

  <size  njmax="15000" nconmax="1000"  nkey="1"/>     
  
  <extension>
      <plugin plugin="mujoco.elasticity.cable"/>
  </extension>

  <default>
    <geom friction=".01"/>
  </default>



  <asset>
    <hfield name="terrain" ncol="201" nrow="201" size="30 30 5 .1"/>
  </asset>


  <worldbody>

        <light diffuse=".5 .5 .5" pos="0 0 10" dir="0 0 -1" />

        <geom type="plane" size="0.4 0.4 0.05" rgba="1 1 1 1" />

        <geom type="plane" pos="0 0 -1" size="10 10 10" rgba="0.5 0.5 0.5 1" />

        <site name="test" pos="-0.05 -0.05 0.06"  size="0.008" rgba="0.8 .9 0 1" />

        <body name="Extrusion">
            <geom type="box" pos="0 0.4 0"   euler="0 0 0" mass="0.2" size="0.4 0.05 0.05" rgba="0.5 0.5 0.5 1"/>
            <geom type="box" pos="0 -0.4 0"  euler="0 0 0" mass="0.2" size="0.4 0.05 0.05" rgba="0.5 0.5 0.5 1"/>
            <geom type="box" pos="0.4 0 0"   euler="0 0 0" mass="0.2" size="0.05 0.4 0.05" rgba="0.5 0.5 0.5 1"/>
            <geom type="box" pos="-0.4 0 0"  euler="0 0 0" mass="0.2" size="0.05 0.4 0.05" rgba="0.5 0.5 0.5 1"/>
        </body>

        <body name="slider_M1" pos="-0.4 0.15 0.06" euler="0 0 135">
            <inertial pos="0 0 0" quat="0 0 0 1" mass="0.1" diaginertia="0.1 0.1 0.1" />
            <geom type = "box" size=".01 0.01 0.01" rgba="1 1 1 1"/>
            <site name="M1" pos="0 0 0"  size="0.008" rgba="0.8 .9 0 1" />
            <joint type="slide" axis="1 0 0" damping="0.1" armature="0.00001" name="slider_joint1"/>
       </body>

       <body name="slider_M3" pos="-0.4 -0.15 0.06" euler="0 0 -135">
            <inertial pos="0 0 0" quat="0 0 0 1" mass="0.1" diaginertia="0.1 0.1 0.1" />
            <geom type = "box" size=".01 0.01 0.01" rgba="1 1 1 1" />
            <site name="M3" pos="0 0 0"  size="0.008" rgba="0.8 .9 0 1" />
            <joint type="slide" axis="1 0 0" damping="0.1" armature="0.00001"  name="slider_joint3"/>
        </body>

       <body name="slider_M2" pos="-0.4 0.3 0.06" euler="0 0 135">
            <inertial pos="0 0 0" quat="0 0 0 1" mass="0.1" diaginertia="0.1 0.1 0.1" />
            <geom type = "box" size=".01 0.01 0.01" rgba="1 1 1 1" />
            <site name="M2" pos="0 0 0"  size="0.008" rgba="0.8 .9 0 1" />
            <joint type="slide" axis="1 0 0" damping="0.1"  armature="0.00001" name="slider_joint2"/>
       </body>

       <body name="slider_M4" pos="-0.4 -0.3 0.06" euler="0 0 -135">
            <inertial pos="0 0 0" quat="0 0 0 1" mass="0.1" diaginertia="0.1 0.1 0.1" />
            <geom type = "box" size=".01 0.01 0.01" rgba="1 1 1 1" />
            <site name="M4" pos="0 0 0"  size="0.008" rgba="0.8 .9 0 1" />
            <joint type="slide" axis="1 0 0" damping="0.1" armature="0.00001" name="slider_joint4"/>
        </body>
     
        <body name = "Link1" pos="-0.3 0 0.06">
            <geom name="1st Link" type="box" pos="0 0 0" mass="1" size=".1 0.02 0.01" />
            <joint name="1st Revolute Joint" pos="-0.1 0 0" type="hinge" axis="0 0 1" limited="true" range="-120 120" frictionloss = "0.2"  damping="0.1" />
            <geom name="1st Revolute Joint" type="cylinder" pos="-0.1 0 0"  size=".02 .01" rgba="0 .9 0 1"  mass="0" group="1" />
            <geom name="offset1" type="box" pos="0 0 0" size=".01 0.05 0.01" />
            <site name="l1" pos="0  0.05 0"  size="0.005" rgba="0.8 .9 0 1" />
            <site name="l2" pos="0 -0.05 0"  size="0.005" rgba="0.8 .9 0 1" />  
            <site name="1st Joint" pos="-0.1 0 0"  size="0.0008" rgba="0.8 .9 0 1" />  


            <body name="DB0" pos="0 0.05 0" euler="0 0 135">
                <composite type = "cable" curve="s 0 0" count="10 1 1"  offset="0 0 0" prefix="D" size="0.1414" initial="ball">
                    <plugin plugin="mujoco.elasticity.cable"/>
                    <joint kind="main" damping="1" />
                    <geom type="capsule" size="0.004" rgba=".8 .2 .1 1"/>
                </composite> 
            </body> 
            <body name="CB0" pos="0 -0.05 0" euler="0 0 -135">
                <composite type="cable" count="10 1 1" curve="s 0 0"  offset="0 0 0" prefix="C" size="0.1414" initial="ball" >
                    <plugin plugin="mujoco.elasticity.cable"/>
                    <joint kind="main" damping="1" />
                    <geom type="capsule" size="0.004" rgba=".8 .2 .1 1"/>
                </composite>
            </body> 

            <body pos="0.2 0 0">
                <geom name="2nd Link" type="box" pos="0 0 0" mass="1"  size=".1 0.02 .01" rgba="0.9 .9 0 1" />
                <joint name="2nd Revolute Joint" pos="-0.1 0 0" type="hinge" axis="0 0 1" limited="true" range="-120 120" frictionloss = "0.2"  damping="0.1" />
                <site name=" 2nd Joint " pos="-0.1 0 0"  size="0.0008" rgba="0.8 .9 0 1" />  
                <geom name="2nd Revolute Joint" type="cylinder" pos="-0.1 0 0"  size=".02 .01" rgba="0 0 0 1"  mass="0" group="1"/>
                <geom name="offset2" type="box" pos="0 0 0" size=".01 0.05 0.01" />
                <site name="l3" pos="0 0.05 0"  size="0.005" rgba="0 .9 0 1" />
                <site name="l4" pos="0 -0.05 0" size="0.005" rgba="0 .9 0 1" />  
                <site name="EE" pos="0.1 0 0" size="0.01" rgba="0 0 0 1" />  
                
                <body name="EB0" pos="0 0.05 0" euler="0 0 140.19">
                    <composite type="cable" count="30 1 1"  curve="s 0 0" offset="0 0 0" prefix="E"  size="0.3905" initial="ball">
                        <plugin plugin="mujoco.elasticity.cable"/>
                        <joint kind="main" damping="1" />
                        <geom type="capsule" size="0.004" rgba=".8 .2 .1 1"/>
                    </composite>
                </body> 
         
                <body name="FB0" pos="0 -0.05 0" euler="0 0 -140.19">
                    <composite type="cable" count="30 1 1" curve="s 0 0"  offset="0 0 0" prefix="F"  size="0.3905" initial="ball" >
                        <plugin plugin="mujoco.elasticity.cable"/>
                        <joint kind="main" damping="1" />
                        <geom type="capsule" size="0.004 " rgba=".8 .2 .1 1"/>
                    </composite>
                </body> 
        
            </body> 
        </body>
  </worldbody>

    <equality>
      <connect name="M1_boundary"  body1="DB_last"  body2="slider_M1" active = "true" anchor="0 0 0" /> #acnhor point is specified according to local coordinate of first body 
      <connect name="M3_boundary"  body1="CB_last"  body2="slider_M3" active = "true" anchor="0 0 0" />
      <connect name="M2_boundary"  body1="EB_last"  body2="slider_M2" active = "true" anchor="0 0 0" />
      <connect name="M4_boundary"  body1="FB_last"  body2="slider_M4" active = "true" anchor="0 0 0" />
    </equality>



    <actuator>   
        <position name="x_axis1"     joint="slider_joint1"    ctrlrange="-0.2 0.2"  kp="50"/>
        <position name="x_axis2"     joint="slider_joint2"    ctrlrange="-0.2 0.2"  kp="50"/>  
        <position name="x_axis3"     joint="slider_joint3"    ctrlrange="-0.2 0.2"  kp="50"/>
        <position name="x_axis4"     joint="slider_joint4"    ctrlrange="-0.2 0.2"  kp="50"/> 
        <velocity name="x_dotaxis1"  joint="slider_joint1"    ctrlrange="-0.2 0.2"  kv="5"/>
        <velocity name="x_dotaxis2"  joint="slider_joint2"    ctrlrange="-0.2 0.2"  kv="5"/>
        <velocity name="x_dotaxis3"  joint="slider_joint3"    ctrlrange="-0.2 0.2"  kv="5"/>
        <velocity name="x_dotaxis4"  joint="slider_joint4"    ctrlrange="-0.2 0.2"  kv="5"/>
    </actuator>   


    </sensor>

</mujoco>`

**Comment by yuvaltassa** [comment]
The code block is empty?
Closing.
